File: c/tree_02.c
// 二叉树的链表实现
// 2020.4.6
#include<stdio.h>
#include<malloc.h>
#include<time.h>
#include<stdlib.h>


#define N 10

typedef int data_type;
typedef struct TreeNode {
    data_type data;
    struct TreeNode *lchild;
    struct TreeNode *rchild;
}tree_node;
typedef tree_node *btree;

btree create_tree(btree root, data_type data)
{
    btree new_node = (btree)malloc(sizeof(tree_node));
    if (new_node != NULL) {
        new_node->data = data;
        new_node->lchild = NULL;
        new_node->rchild = NULL;
        if (root == NULL) {
            root = new_node;
            return root;
        }
        else {
            btree ptr = root, current;
            while (ptr != NULL) {
                current = ptr;
                if (ptr->data > data)
                    ptr = ptr->lchild;
                else
                    ptr = ptr->rchild;
            }
            if (current->data > data)
                current->lchild = new_node;
            else 
                current->rchild = new_node;
        }
        return root;
    }
    else {
        printf("Failed to create new node!\n");
        exit(-1);
    }
}

void free_tree(btree node)
{
    if (node != NULL) {
        free_tree(node->lchild);
        free_tree(node->rchild);
        printf("%d\t", node->data);
        free(node);
        node = NULL;
    }
}

void main(void)
{
    data_type array[N];
    btree root = NULL;
    srand((unsigned int)time(NULL));
    for (int i = 0; i < N; i++) {
        array[i] = rand() % 20;
        printf("%d\t", array[i]);
        root = create_tree(root, array[i]);
    }
    printf("\n");
    btree ptr = root;
    printf("left subtree:\n");
    ptr = root;
    while (ptr != NULL) {
        printf("%d\t", ptr->data);
        ptr = ptr->lchild;
    }
    printf("\n");
    free_tree(root);
    printf("\n");
}

File: python/tree_02.py
# 树的链表表示法
class tree:
    def __init__(self):
        self.data = 0
        self.left = None
        self.right = None


def create_tree(root, val):
    newnode = tree()
    newnode.data = val
    newnode.left = None
    newnode.right = None
    if root == None:
        root = newnode
        return root
    else:
        current = root
        while current != None:
            backup = current
            if current.data > val:
                current = current.left
            else:
                current = current.right
        if backup.data > val:
            backup.left = newnode
        else:
            backup.right = newnode
    return root

data = [5, 6, 24, 8, 12, 3, 17, 1, 9]
ptr = None
root = None
for i in range(9):
    ptr = create_tree(ptr, data[i])
print('left subtree:')
root = ptr.left
while root != None:
    print('%d' % root.data)
    root = root.left
print('right subtree:')
root = ptr.right
while root != None:
    print('%d' % root.data)
    root = root.right
